pkg/uhttp/gocache.go

```diff
@@ -0,0 +1,170 @@
+package uhttp
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"io"
+	"net/http"
+	"net/http/httputil"
+	"time"
+
+	bigcache "github.com/allegro/bigcache/v3"
+	"github.com/grpc-ecosystem/go-grpc-middleware/logging/zap/ctxzap"
+	"go.uber.org/zap"
+)
+
+type GoCache struct {
+	ttl         time.Duration
+	rootLibrary *bigcache.BigCache
+}
+
+func NewGoCache(ctx context.Context, ttl int32, cacheMaxSize int, isLogLevel bool) (GoCache, error) {
+	l := ctxzap.Extract(ctx)
+	config := bigcache.Config{
+		// number of shards (must be a power of 2)
+		Shards: 1024,
+
+		// time after which entry can be evicted
+		LifeWindow: time.Duration(ttl) * time.Second,
+
+		// Interval between removing expired entries (clean up).
+		// If set to <= 0 then no action is performed.
+		// Setting to < 1 second is counterproductive — bigcache has a one second resolution.
+		CleanWindow: 5 * time.Minute,
+
+		// rps * lifeWindow, used only in initial memory allocation
+		MaxEntriesInWindow: 1000 * 10 * 60,
+
+		// max entry size in bytes, used only in initial memory allocation
+		MaxEntrySize: 500,
+
+		// prints information about additional memory allocation
+		Verbose: isLogLevel,
+
+		// cache will not allocate more memory than this limit, value in MB
+		// if value is reached then the oldest entries can be overridden for the new ones
+		// 0 value means no size limit
+		// Default value "GB eq 2048MB
+		HardMaxCacheSize: cacheMaxSize, // BATON_CACHE_MAX_SIZE
+
+		// callback fired when the oldest entry is removed because of its expiration time or no space left
+		// for the new entry, or because delete was called. A bitmask representing the reason will be returned.
+		// Default value is nil which means no callback and it prevents from unwrapping the oldest entry.
+		OnRemove: nil,
+
+		// OnRemoveWithReason is a callback fired when the oldest entry is removed because of its expiration time or no space left
+		// for the new entry, or because delete was called. A constant representing the reason will be passed through.
+		// Default value is nil which means no callback and it prevents from unwrapping the oldest entry.
+		// Ignored if OnRemove is specified.
+		OnRemoveWithReason: nil,
+	}
+	cache, initErr := bigcache.New(ctx, config)
+	if initErr != nil {
+		l.Error("in-memory cache error", zap.Any("NewGoCache", initErr))
+		return GoCache{}, initErr
+	}
+
+	gc := GoCache{
+		ttl:         config.LifeWindow,
+		rootLibrary: cache,
+	}
+
+	return gc, nil
+}
+
+func GetCacheKey(req *http.Request) string {
+	return req.URL.String()
+}
+
+func CopyResponse(resp *http.Response) *http.Response {
```

Comment:
Is the plan to use this? I don't see it called anywhere.

Reply:
Fixed